rust/template/distributed_datalog/src/zookeeper.rs

```diff
@@ -0,0 +1,142 @@
+//! A module providing ZooKeeper related functionality for `d3log`.
+
+#[cfg(test)]
+use std::env::var_os;
+use std::time::Duration;
+
+use serde::de::DeserializeOwned;
+use serde_json::from_slice;
+
+use zookeeper::Acl;
+use zookeeper::CreateMode;
+use zookeeper::ZkError;
+
+use crate::schema::Members;
+use crate::schema::SysCfg;
+use crate::ReadConfig;
+use crate::ReadMembers;
+
+pub use zookeeper::WatchedEvent;
+pub use zookeeper::Watcher;
+pub use zookeeper::ZooKeeper;
+
+#[cfg(test)]
+const ENDPOINTS_ENV: &str = "ZOOKEEPER_ENDPOINTS";
+
+const D3LOG_KEY: &str = "/d3log";
+const CONFIG_KEY: &str = "/d3log/config";
+const MEMBERS_KEY: &str = "/d3log/members";
+
+/// Connect to a ZooKeeper instance comprised of the given set of servers.
+///
+/// Note that each server address should contain the port the
+/// service is listening on.
+pub fn connect<I, S, W>(mut servers: I, watcher: W) -> Result<ZooKeeper, String>
+where
+    I: Iterator<Item = S>,
+    S: AsRef<str>,
+    W: Watcher + 'static,
+{
+    let first = servers
+        .next()
+        .ok_or_else(|| "failed to connect to ZooKeeper: no servers supplied")?
+        .as_ref()
+        .to_string();
+    let servers = servers.fold(first, |x, y| x + "," + y.as_ref());
```

Comment:
Would it be simple to use `join()` instead?
https://doc.rust-lang.org/beta/std/primitive.slice.html#examples-71

Reply:
The reason I decided against it is because `join` seemingly requires a slice, which we don't have. So we'd need to `collect` into a `Vec` first to make this work, which seems wasteful.